prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,213 @@
+func NewManager() Manager {
+	return single
+}
+
+// Manager knows how to run tasks with graceful termination on interrupts
+type Manager interface {
+	ListenAndServe(server *http.Server, gracePeriod time.Duration)
+	ListenAndServeTLS(server *http.Server, certFile, keyFile string, gracePeriod time.Duration)
+	Run(work func(ctx context.Context))
+	Tick(work func(), interval time.Duration)
+	Context() context.Context
+	WaitForGracefulShutdown()
+}
+
+type manager struct {
+	// only one signal handler should be installed, so we use a cond to
+	// broadcast to workers that an interrupt has occurred
+	c *sync.Cond
+	// we record whether we've broadcast in the past
+	seenSignal bool
+	// we want to ensure that all registered servers and workers get a
+	// change to gracefully shut down
+	wg sync.WaitGroup
+}
+
+// handleInterrupt turns an interrupt into a broadcast for our condition.
+// This must be called _first_ before any work is registered with the
+// manager, or there will be a deadlock.
+func (m *manager) handleInterrupt() {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM, syscall.SIGABRT)
+	s := <-sig
+	logrus.WithField("signal", s).Info("Received signal.")
+	m.seenSignal = true
+	m.c.Broadcast()
+}
+
+// WaitForGracefulShutdown waits until all registered servers and workers
+// have had time to gracefully shut down, or times out. This function is
+// blocking.
+func (m *manager) WaitForGracefulShutdown() {
+	finished := make(chan struct{})
+	go func() {
+		defer close(finished)
+		if !m.seenSignal {
+			m.c.Wait()
+		}
+	}()
+	select {
+	case <-finished:
+		logrus.Info("All workers gracefully terminated, exiting.")
+	case <-time.After(1 * time.Minute):
+		logrus.Warn("Timed out waiting for workers to gracefully terminate, exiting.")
+	}
+}
+
+// Context returns a context that stays is cancelled when an interrupt hits.
+// Using this context is a weak guarantee that your work will finish before
+// process exit as callers cannot signal that they are finished. Prefer to use
+// Run().
+func (m *manager) Context() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+
+		m.c.L.Lock()
+		if !m.seenSignal {
+			m.c.Wait()
+		}
+		cancel()
+		m.c.L.Unlock()
+	}()
+
+	return ctx
+}
+
+// Run will do work until an interrupt is received, then signal the
+// worker. This function is not blocking. Callers are expected to exit
+// only after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down. This is preferable to getting the raw Context
+// as we can ensure that the work is finished before releasing our share
+// of the wait group on shutdown.
+func (m *manager) Run(work func(ctx context.Context)) {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		work(ctx)
+	}()
+
+	m.c.L.Lock()
+	if !m.seenSignal {
+		m.c.Wait()
+	}
+	cancel()
```

Comment:
nit: These `cancel()` calls could be made without holding the lock. I'm pretty sure they're very fast though so it probably doesn't actually matter.